zanata-war/src/main/java/org/zanata/service/impl/TranslationStateCacheImpl.java

```diff
@@ -191,18 +191,22 @@ public Boolean textFlowTargetHasWarningOrError(Long targetId,
     // TODO why not @Observe the event directly?
     @Override
     public void textFlowStateUpdated(TextFlowTargetStateEvent event) {
-        DocumentLocaleKey key =
-                new DocumentLocaleKey(event.getDocumentId(),
-                        event.getLocaleId());
+        Long documentId = event.getKey().getDocumentId();
+        LocaleId localeId = event.getKey().getLocaleId();
 
         //invalidate document statistic cache
-        clearDocumentStatistics(event.getDocumentId(), event.getLocaleId());
+        clearDocumentStatistics(documentId, localeId);
 
-        // update document status information
-        updateDocStatusCache(key, event.getTextFlowTargetId());
+        for(TextFlowTargetStateEvent.TextFlowTargetState state: event.getStates()) {
+            Long tftId = state.getTextFlowTargetId();
 
-        // invalidate target validation
-        targetValidationCache.remove(event.getTextFlowTargetId());
+            // update document status information
+            updateDocStatusCache(new DocumentLocaleKey(documentId,
+                localeId), tftId);
```

Comment:
So why not pass in the last TFT details from the ImmutableList, and use that to update docStatusCache? All the TFTs should have the same timestamp and account if they're part of a batch, but using the last one's metadata should give the desired effect.